pkg/controller/certificates/controller.go

```diff
@@ -121,33 +125,48 @@ func NewCertificateController(kubeClient clientset.Interface, syncPeriod time.Du
 // Run the main goroutine responsible for watching and syncing jobs.
 func (cc *CertificateController) Run(workers int, stopCh <-chan struct{}) {
 	defer utilruntime.HandleCrash()
+	defer cc.queue.ShutDown()
+
 	go cc.csrController.Run(stopCh)
+
+	if !framework.WaitForCacheSync(stopCh, cc.csrStoreSynced) {
+		return
+	}
+
 	glog.Infof("Starting certificate controller manager")
 	for i := 0; i < workers; i++ {
 		go wait.Until(cc.worker, time.Second, stopCh)
 	}
 	<-stopCh
 	glog.Infof("Shutting down certificate controller")
-	cc.queue.ShutDown()
 }
 
 // worker runs a thread that dequeues CSRs, handles them, and marks them done.
 func (cc *CertificateController) worker() {
-	for {
-		func() {
-			key, quit := cc.queue.Get()
-			if quit {
-				return
-			}
-			defer cc.queue.Done(key)
-			err := cc.syncHandler(key.(string))
-			if err != nil {
-				glog.Errorf("Error syncing CSR: %v", err)
-			}
-		}()
+	defer glog.Infof("certificate controller worker shutting down")
```

Comment:
> Hmm, what about "certificate controller work queue is quit"? Or, print nothing?

I'd print nothing.  Any failure here will trigger either a `utilruntime.HandleError` or `utilruntime.HandleCrash`.  The clean shutdown is handled in `Run`